core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2375,25 +2375,25 @@ private void checkStorageTableNotRedirected(QualifiedObjectName source)
 
         private void analyzeFiltersAndMasks(Table table, QualifiedObjectName name, RelationType relationType, Scope accessControlScope)
         {
-            ImmutableList.Builder<ColumnSchema> columnSchemaBuilder = ImmutableList.builder();
+            ImmutableMap.Builder<String, ColumnSchema> columnSchemaBuilder = ImmutableMap.builder();
             for (int index = 0; index < relationType.getAllFieldCount(); index++) {
                 Field field = relationType.getFieldByIndex(index);
-                field.getName().ifPresent(fieldName -> columnSchemaBuilder.add(ColumnSchema.builder()
+                field.getName().ifPresent(fieldName -> columnSchemaBuilder.put(fieldName, ColumnSchema.builder()
                         .setName(fieldName)
                         .setType(field.getType())
                         .setHidden(field.isHidden())
                         .build()));
             }
-            List<ColumnSchema> columnSchemas = columnSchemaBuilder.build();
+            Map<String, ColumnSchema> columnSchemas = columnSchemaBuilder.buildOrThrow();
 
-            Map<ColumnSchema, ViewExpression> masks = accessControl.getColumnMasks(session.toSecurityContext(), name, columnSchemas);
+            Map<ColumnSchema, ViewExpression> masks = accessControl.getColumnMasks(session.toSecurityContext(), name, ImmutableList.copyOf(columnSchemas.values()));
 
-            for (ColumnSchema columnSchema : columnSchemas) {
-                Optional.ofNullable(masks.get(columnSchema)).ifPresent(mask -> {
-                    if (checkCanSelectFromColumn(name, columnSchema.getName())) {
+            if (checkCanSelectFromColumns(name, columnSchemas.keySet())) {
```

Comment:
We could move the call to `accessControl.getColumnMasks` into this `if` block so the column masks are not fetched if they are not going to be processed, wdyt?

A similar logic may be applicable to the row filters; I imagine row filters aren't useful if the session can't select from column. I am not very familiar with how trino analyses row filters so please double-check.

Reply:
> We could move the call to `accessControl.getColumnMasks` into this `if` block so the column masks are not fetched if they are not going to be processed, wdyt?

I had to rework the code a bit, because I was not aware of some special behavior with `hide-inaccessible-columns` (thankfully Trino has excellent test coverage), so now it's not beneficial anymore. But good point.

> A similar logic may be applicable to the row filters; I imagine row filters aren't useful if the session can't select from column. I am not very familiar with how trino analyses row filters so please double-check.

Row filters are on a table level, so that's a little different, and the filtering expression is analyzed as part of the `WHERE` clause. We could imagine a bulk `getRowFilters`, though - just as there are tables with thousands of columns, there are some crazy queries that reference thousands of tables.

Reply:
> Row filters are on a table level, so that's a little different, and the filtering expression is analyzed as part of the `WHERE` clause.

On the other hand I may be out of my depth here as well :)

Reply:
> On the other hand I may be out of my depth here as well :)

It seems I totally am - that method is called before the analyzer knows which columns are referenced by the query, so we get all columns from the table. That access check was introduced to avoid confusing failures when we analyze masks on columns which are not involved in the query at all.